[wallet] Restore ability to list incoming transactions by label

ryanofsky opened the pull request:
This change partially reverts #13075 and #14023.

Fixes #14382

ryanofsky committed:
[wallet] Restore ability to list incoming transactions by label

This change partially reverts #13075 and #14023.

Fixes #14382

MarcoFalke commented:
Is this for backport to 0.17.1? While not necessary, I believe it would simplify upgrade from 0.17.1+ to 0.18.0, because 0.17.1 could be used without `-deprecatedrpc` as workaround.

ryanofsky commented:
I think it'd be reasonable to backport this, since it's a small change that restores a removed feature. But I think it won't cherry-pick cleanly because #14023 was merged since the 0.17 branch. @Saicere originally reported the issue so they might be able to say whether there's a benefit to backporting.

Saicere commented:
> @Saicere originally reported the issue so they might be able to say whether there's a benefit to backporting.

As far as I am concerned, it's not really a big deal whether or not it's backported to the 0.17 branch as long as it's in before accounts are removed entirely. If you were using accounts in your application, you are probably running 0.17 with `-deprecatedrpc=accounts` anyway while RPC-facing code is updated.

kristapsk commented:
ACK 8fcb76508493cd09d4895e7f3ad11cc5b945056e

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1375,8 +1375,9 @@ static void MaybePushAddress(UniValue & entry, const CTxDestination &dest)
+ * @param  filter_label Optional label string to filter incoming transactions.
  */
-static void ListTransactions(CWallet* const pwallet, const CWalletTx& wtx, int nMinDepth, bool fLong, UniValue& ret, const isminefilter& filter) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+static void ListTransactions(CWallet* const pwallet, const CWalletTx& wtx, int nMinDepth, bool fLong, UniValue& ret, const isminefilter& filter, const std::string* filter_label) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
```
>  @param  filter_label Optional label string to filter incoming transactions.

nit, `filter_label = nullptr`. Also, makes the diff smaller

jnewbery replied:
I agree that default arguments should be used sparingly.

promag replied:
:+1:

promag reviewed (commented):
Concept ACK.

jnewbery commented:
Concept ACK. For completeness, this functionality was hidden behind a deprecation switch in #12953 (based on #11497). The functionality was marked deprecated in #5575.

Apologies - it was my oversight that this functionality got dropped. I'm happy to backport this to V0.17 once this gets merged. I aim to review this week.

jnewbery reviewed (commented):
Tested ACK 8fcb76508493cd09d4895e7f3ad11cc5b945056e with a couple of nits.

@MarcoFalke - I'll go ahead and open a PR for backporting. A couple of questions:
- does the backport need release notes?
- if we add release notes for the backport, do we also need release notes here?

jnewbery commented:
Backport is #14441. PR numbering was totally intentional.

ryanofsky reviewed (commented):
Added 1 commit 8fcb76508493cd09d4895e7f3ad11cc5b945056e -> 665d00fce4bc53a9ea9dc5a7d457d586af357bf3 ([pr/list-label.1](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.1) -> [pr/list-label.2](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/list-label.1...pr/list-label.2)) with suggested fixes.
Squashed 665d00fce4bc53a9ea9dc5a7d457d586af357bf3 -> fa7fae5c3612cf77c91b92a39c14e73478424e85 ([pr/list-label.2](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.2) -> [pr/list-label.3](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.3))
Updated fa7fae5c3612cf77c91b92a39c14e73478424e85 -> 4deba4ced9fc4edd1a40837c083080b6ed4c8163 ([pr/list-label.3](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.3) -> [pr/list-label.4](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.4)) with release notes changes to reflect the intent to backport this to 0.17.1

jnewbery commented:
The release note should be removed from this PR (assuming #14441 is merged)

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1460,10 +1464,12 @@ UniValue listtransactions(const JSONRPCRequest& request)
     if (request.fHelp || request.params.size() > 4)
         throw std::runtime_error(
-            "listtransactions (dummy count skip include_watchonly)\n"
+            "listtransactions (label count skip include_watchonly)\n"
```
nit, fix consistency by adding spaces after `(` and before `)`?

ryanofsky commented:
Updated 4deba4ced9fc4edd1a40837c083080b6ed4c8163 -> 7cbe74f1524fc8621eb339d5eddd9530826c8461 ([pr/list-label.4](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.4) -> [pr/list-label.5](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.5)) removing release notes and changing whitespace as suggested.

sipa commented:
Very nitty comment; it feels like this functionality is closer to `listreceivedbylabel` than to `listtransactions` (as the latter is about all transactions which affect the balance of the wallet - and formerly account) rather than just incoming payments. Would it make sense to add it as an extra option to `listreceivedbylabel` (verbose or so), or a separate RPC, instead?

Saicere commented:
> Would it make sense to add it as an extra option to `listreceivedbylabel` (verbose or so), or a separate RPC, instead?

`listreceivedbylabel` doesn't take a label argument or list transactions in any way, it just lists total received coins per label, as per [the code](https://github.com/bitcoin/bitcoin/blob/be992701b018f256db6d64786624be4cb60d8975/src/wallet/rpcwallet.cpp#L1318).

jnewbery commented:
ACK 7cbe74f1524fc8621eb339d5eddd9530826c8461. Only change since 4deba4c is removing release notes and changing whitespace.

ryanofsky commented:
> Would it make sense to add it as an extra option to listreceivedbylabel (verbose or so), or a separate RPC, instead?

It'd be good to improve listreceivedbylabel. But filtering by label in `listtransactions` used to work previously and I think was just removed by accident.

You could maybe say that `listtransactions` shouldn't return labels or filter by label at all, since outgoing transactions won't have labels. But I don't think it makes sense to keep returning labels and only remove the ability to filter them.

MeshCollider commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1375,8 +1375,9 @@ static void MaybePushAddress(UniValue & entry, const CTxDestination &dest)
  * @param  fLong      Whether to include the JSON version of the transaction.
  * @param  ret        The UniValue into which the result is stored.
  * @param  filter     The "is mine" filter bool.
```
IMO this should be renamed to avoid confusion, to something like `filter_ismine`

meshcollider reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/14411/commits/7cbe74f1524fc8621eb339d5eddd9530826c8461

ryanofsky committed:
Rename ListTransactions filter variable

Suggested by MeshCollider <dobsonsa68@gmail.com> in
https://github.com/bitcoin/bitcoin/pull/14411#discussion_r232134086

ryanofsky commented:
Rebased 7cbe74f1524fc8621eb339d5eddd9530826c8461 -> 671d8ee370de4824f4ab30b904bf59514f5e52e5 ([pr/list-label.5](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.5) -> [pr/list-label.6](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.6)) due to conflict with #14437. Also renamed filter variable as suggested by MeshCollider.

MarcoFalke commented:
Unless I am mistaken this has been "backported" in 5150accdd2a7c7f0edf964d56bd7d34b5f740cdc. So this must go into 0.18.0 instead.
(assigned new milestone)

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1352,10 +1356,12 @@ UniValue listtransactions(const JSONRPCRequest& request)
     if (request.fHelp || request.params.size() > 4)
         throw std::runtime_error(
-            "listtransactions (dummy count skip include_watchonly)\n"
+            "listtransactions ( label count skip include_watchonly )\n"
```
From the documentation below, label is a string, so should be surrounded by quotation marks in the oneline doc. 

```suggestion
            "listtransactions ( \"label\" count skip include_watchonly )\n"
```

ryanofsky reviewed (commented):
Rebased 671d8ee370de4824f4ab30b904bf59514f5e52e5 -> da427dbd48ae8d12a2a79a7514a813e78064fe52 ([pr/list-label.6](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.6) -> [pr/list-label.7](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.7)) due to conflict with #14720

MarcoFalke commented:
ACK da427dbd48ae8d12a2a79a7514a813e78064fe52 (Only checked that the rebase was done properly from the last time people reviewed this. Also checked that the tests are the same as in the backport 5150accdd2a7c7f0edf964d56bd7d34b5f740cdc, didn't look at the code)

jnewbery commented:
utACK da427dbd48ae8d12a2a79a7514a813e78064fe52
